polynote-server/src/main/scala/polynote/server/repository/NotebookRepository.scala

```diff
@@ -500,35 +503,63 @@ class FileBasedRepository(
           case None            => ZIO.unit
         }
 
+      val initWAL = get.flatMap {
+        notebook => Message.encode[Task](notebook.withoutResults).flatMap(bytes => effectBlocking(wal.write(bytes.toByteArray)))
+      }.catchAll {
+        err => Logging.error("Unable to initialize WAL", err)
+      }
+
+      def writeWAL(update: NotebookUpdate) =
+        Message.encode[Task](update).flatMap {
+          message => effectBlocking {
+            wal.synchronized {
+              wal.write(message.toByteArray)
+            }
+          }
+        }.catchAll {
+          err => Logging.error("Unable to write update to WAL (this WAL segment will be useless)", err) // TODO: improve this
+        }
+
+      val syncWAL = effectBlocking(wal.synchronized(wal.flush())).catchAll {
+        err => Logging.error("Unable to sync WAL (this WAL segment will be useless)", err)
+      }
+
+      val closeWAL = effectBlocking(wal.close()).catchAll {
+        err => Logging.error("Failed to close WAL", err)
+      }
+
       // process all the submitted updates into the ref in order
       val processPending = pending.take.flatMap {
-        case Take.Value((update, completerOpt)) =>
-          (doUpdate(update) >>= notifyCompleter(completerOpt)).asSomeError
+        case Take.Value((update, completerOpt)) => (writeWAL(update).forkDaemon &> doUpdate(update) >>= notifyCompleter(completerOpt)).asSomeError
         case Take.End                           => pending.shutdown *> closed.succeed(()).unit <* ZIO.fail(None)
```

Comment:
FYI, I'm seeing this warning here: 
```[warn] /Users/jindig/projects/polynote/polynote-server/src/main/scala/polynote/server/repository/NotebookRepository.scala:532:49: match may not be exhaustive.
[warn] It would fail on the following input: Fail(_)
[warn]       val processPending = pending.take.flatMap {
[warn]                                                 ^
[warn] one warning found
```

Reply:
Yeah, I guess Scala isn't smart enough to realize that `Take[Nothing, ...]` can't be a `Take.Fail`. I guess I should add a case for it even though it's impossible.